llvm/lib/CodeGen/LLVMTargetMachine.cpp

```diff
@@ -37,6 +37,11 @@ static cl::opt<bool>
     EnableTrapUnreachable("trap-unreachable", cl::Hidden,
                           cl::desc("Enable generating trap for unreachable"));
 
+static cl::opt<bool> EnableNoTrapAfterNoreturn(
+    "no-trap-after-noreturn", cl::Hidden,
+    cl::desc("Do not emit a trap instruction for 'unreachable' IR instructions "
+             "after noreturn calls, even if --trap-unreachable is set."));
```

Comment:
Why do we need this flag? And why do we need to test wasm code with this flag, given that we set it to false?

Reply:
And given that `NoTrapAfterNoreturn` is by default set to false and we don't override it, we've been setting this to false anyway... Why do we need to newly 'ensure' it? https://github.com/llvm/llvm-project/blob/2cab996192cf143d10e3381fcefa75e270cc7ddb/llvm/include/llvm/Target/TargetOptions.h#L137

Reply:
I added the flag because I couldn't test how the feature effects wasm output without the ability to enable it from the command line.

Being able to test NoTrapAfterNoreturn's behaviour prevents this bug from regressing. In the future, people may want to stop forcing TrapUnreachable to true and NoTrapAfterNoreturn to false, if the wasm backend improves and it stops being necessary for correctness. It seems to me that if a feature exists in llvm and is able to cause bugs, we should be able to test it.

While the NoTrapAfterNoreturn option is false by default, this can be overridden by frontends like clang or the Rust compiler. [Like this change to the Rust compiler](https://github.com/rust-lang/rust/pull/110494/files#diff-eac3d5ef63f39b7914ae9015f31b8d87353ea8fbcd5b02a7335dd86c557e7625), which is the original motivation behind this pull request (expand the code down and look for the line with `createTargetMachine()` for an idea of how it works). This is actually the purpose of the TargetOptions class, to pass (command line) options from the frontend to llvm.

The wasm backend should still produce valid code even if it receives the NoTrapAfterNoreturn option, is my feeling.

Reply:
As you correctly pointed out in https://github.com/llvm/llvm-project/pull/65876#issuecomment-1714681125, if we turn `NoTrapAfterNoreturn` on, our backend produces invalid code. I don't think our tests should test if we (correctly) produce invalid code when a command line option is turned on.

And why does the rust code need to turn that option on? I scrolled down as you said but failed to find out why..

> In the future, people may want to stop forcing TrapUnreachable to true and NoTrapAfterNoreturn to false, if the wasm backend improves and it stops being necessary for correctness.

If we want to test something, I think we should just test the current behavior. Not sure if we need to predict the future scenario and test it in advance.

Reply:
> I don't think our tests should test if we (correctly) produce invalid code when a command line option is turned on.

They don't- the tests I added test that we produce *valid* code, *regardless* of the TargetOptions passed to us. I also disagree that it's correct to produce invalid code depending on a TargetOption. What's the point in emitting code that can't be executed?

I think you may have missed [this change](https://github.com/llvm/llvm-project/pull/65876/files#diff-174c1ba15a00ab323f265d131a24ddd8854ab4ce70133d83ec52972f188f9559R130), which is the most important line and the actual bug fix.

The point is that code external to llvm can pass NoTrapAfterNoreturn as True, just as it can pass TrapUnreachable as False. However, the wasm backend is unusual in that it (ab)uses TrapUnreachable to emit correct code. And so in the WebAssemblyTargetMachine constructor, TrapUnreachable is hard-coded to True. My change is that NoTrapAfterNoreturn must also be hard-coded to False to always emit correct code. This is because NoTrapAfterNoreturn has the effect of suppressing TrapUnreachable after a noreturn call, which can cause invalid wasm.

> And why does the rust code need to turn that option on?

[Rust has had TrapUnreachable on for a while now](https://github.com/rust-lang/rust/pull/45920), because it worked around a few [nasty soundness bugs](https://github.com/rust-lang/rust/issues/28728), and it limits the effect of undesired undefined behaviour in some cases. They don't seem confident that it's safe to turn off yet.
Rust *wants* to turn NoTrapAfterNoreturn on across all llvm backends, because it cuts down on unnecessary trap instructions being emitted after noreturn calls, a common case in Rust code, and should be safe.
You also see [silly double traps](https://godbolt.org/z/1z6vMGavz) when the code is intentionally trying to trap, because llvmir's `@llvm.trap()` is a noreturn function. This is what I was trying to fix up with my wasm peephole optimisation, removing obviously unnecessary double `unreachable`s, to make up for the fact that we can't use NoTrapAfterNoreturn on the wasm backend.

Reply:
> I think you may have missed [this change](https://github.com/llvm/llvm-project/pull/65876/files#diff-174c1ba15a00ab323f265d131a24ddd8854ab4ce70133d83ec52972f188f9559R130), which is the most important line and the actual bug fix.

Why is this a bugfix? It has been false so far anyway.

> The point is that code external to llvm can pass NoTrapAfterNoreturn as True, just as it can pass TrapUnreachable as False. However, the wasm backend is unusual in that it (ab)uses TrapUnreachable to emit correct code. And so in the WebAssemblyTargetMachine constructor, TrapUnreachable is hard-coded to True. My change is that NoTrapAfterNoreturn must also be hard-coded to False to always emit correct code. This is because NoTrapAfterNoreturn has the effect of suppressing TrapUnreachable after a noreturn call, which can cause invalid wasm.
> 
> > And why does the rust code need to turn that option on?
> 
> [Rust has had TrapUnreachable on for a while now](https://github.com/rust-lang/rust/pull/45920), because it worked around a few [nasty soundness bugs](https://github.com/rust-lang/rust/issues/28728), and it limits the effect of undesired undefined behaviour in some cases. They don't seem confident that it's safe to turn off yet. Rust _wants_ to turn NoTrapAfterNoreturn on across all llvm backends, because it cuts down on unnecessary trap instructions being emitted after noreturn calls, a common case in Rust code, and should be safe. You also see [silly double traps](https://godbolt.org/z/1z6vMGavz) when the code is intentionally trying to trap, because llvmir's `@llvm.trap()` is a noreturn function. This is what I was trying to fix up with my wasm peephole optimisation, removing obviously unnecessary double `unreachable`s, to make up for the fact that we can't use NoTrapAfterNoreturn on the wasm backend.

Can you give an example of when it is safe to set `NoTrapAfterNoreturn` on and when it is not? 

Btw the current PR makes setting `--no-trap-after-noreturn` true impossible, because you overrided it in `WebAssemblyTargetMachine.cpp`.

Reply:
> Why is this a bugfix? It has been false so far anyway.

Because it can be passed to us as true by llvm's users (e.g. clang, rustc) because it's part of the public API.

> Can you give an example of when it is safe to set NoTrapAfterNoreturn on and when it is not?

For other backends like x86 or ARM, it should always be safe to use `--trap-unreachable` and `--no-trap-after-noreturn`.

However, the WebAssembly backend is special, because it uses the `TrapUnreachable` option outside of its originally intended purpose.
The WebAssembly backend relies on the fact that `TrapUnreachable` is true to produce correct code, so it forces this option to true. 
However, it *also* relies on `NoTrapAfterNoreturn` being false to produce correct code. The original author of the code didn't realise this, and didn't force it to false. This PR fixes this. 

> Btw the current PR makes setting --no-trap-after-noreturn true impossible, because you overrided it in WebAssemblyTargetMachine.cpp.

Only for the WebAssembly backend (which is kinda the point). For other backends it works as expected.